ls1intum/Artemis
# [Enhancement] Show capped penalty in feedback view

### Checklist
- [x] Server: I updated integration tests (Spring) related to the features (with a high test coverage)
- [x] Client: I updated integration tests (Jest) related to the features (with a high test coverage)

### Motivation and Context



For static code analysis issues, the instructor can specify penalty limits for the exercise as well as for each sca category. It is currently not shown to the students when and how those limits are applied to the score. This might lead to confusion for a student, because solving sca issues might not increase the score when above a penalty limit.

### Description


- Added a property in the sca issue object inside the feedback object to contain the original penalty for the category of the issue.
    - (The credits field already contains the reduced penalty with applied limits)
- In the feedback list, the original penalty is shown now (not affected by limits)
- In the score graph, the 'would-be' deductions without any limits are noted in the legend (only if they are actually different to the applied deductions)
- The legend now displays a tooltip to explain the penalty limits to the students (only if there are any applied limits)

### Steps for Testing


1. Log in to Artemis
2. Create a programming exercise or use an existing one
3. Go to the grading page and configure the sca penalties and especially max-penalties for the categories, such that the limits apply when there are multiple issues in one category (e.g. set both to equal values)
4. Participate as a student and make sure to have multiple sca issues in one category (e.g. introduce several magic number issues)
5. View your feedback and score, check the legend to state the deductions with and without limits and check the tooltip.

### Test Coverage


- result-detail.component.ts 94.81%
- scoreChartPreset.ts 92.73%
- StaticCodeAnalysisService 99%

### Screenshots


 
#### 3 Issues in a category:
<img width="785" alt="Bildschirmfoto 2020-11-24 um 11 35 10" src="https://user-images.githubusercontent.com/13920539/100083010-33b62f00-2e49-11eb-8ad5-8594671df3af.png">

#### The category configuration:
<img width="857" alt="Bildschirmfoto 2020-11-24 um 11 35 17" src="https://user-images.githubusercontent.com/13920539/100083014-34e75c00-2e49-11eb-9432-99585d96afb9.png">

#### The resulting score graph:
<img width="788" alt="Bildschirmfoto 2020-11-24 um 11 35 27" src="https://user-images.githubusercontent.com/13920539/100083017-357ff280-2e49-11eb-9afc-1a91132414fa.png">
